File: packages/integrations/synap-smolagents/synap_smolagents/tools.py
"""Synap memory tools for Smolagents agents.

`create_synap_tools` returns two Smolagents `@tool` functions — `search_memory` and
`store_memory` — that the model can call. Especially idiomatic for `CodeAgent`,
which invokes tools as plain Python calls.

Smolagents tools are synchronous (async `forward` is not supported), so each tool
drives the async Synap SDK through `run_async`. Reads degrade (return a not-found
message); the write raises `SynapIntegrationError` so a failed store is observable.
"""

from __future__ import annotations

import logging
from typing import List

from smolagents import tool
from maximem_synap import MaximemSynapSDK
from synap_integrations_common import run_async, wrap_sdk_errors

logger = logging.getLogger(__name__)


def create_synap_tools(
    sdk: MaximemSynapSDK,
    user_id: str,
    *,
    customer_id: str = "",
) -> List:
    """Build `search_memory` and `store_memory` tools bound to a Synap scope.

    Args:
        sdk: Configured :class:`MaximemSynapSDK`.
        user_id: Synap user scope. **Required.**
        customer_id: Optional customer/org scope; empty means customer-less.

    Returns:
        A list ``[search_memory, store_memory]`` of Smolagents tools.
    """
    if sdk is None:
        raise ValueError("create_synap_tools requires a non-None sdk")
    if not user_id or not str(user_id).strip():
        raise ValueError("create_synap_tools requires a non-empty user_id")

    @tool
    def search_memory(query: str) -> str:
        """Search the user's long-term memory for relevant context.

        Args:
            query: What to look up in the user's memory.
        """
        try:
            response = run_async(
                sdk.fetch(
                    user_id=user_id,
                    customer_id=customer_id or None,
                    search_query=[query],
                    include_conversation_context=False,
                )
            )
        except Exception as exc:  # noqa: BLE001 — read-side graceful degrade
            logger.error(
                "create_synap_tools.search_memory failed user_id=%s error=%s",
                user_id,
                exc,
                exc_info=True,
            )
            return "No relevant memories found."
        return (
            getattr(response, "formatted_context", None)
            or "No relevant memories found."
        )

    @tool
    def store_memory(content: str) -> str:
        """Store a fact or note in the user's long-term memory for future recall.

        Args:
            content: The information to remember.
        """
        with wrap_sdk_errors("smolagents.store_memory", logger, user_id=user_id):
            result = run_async(
                sdk.memories.create(
                    document=content,
                    user_id=user_id,
                    customer_id=customer_id or None,
                )
            )
        return (
            f"Memory stored (ingestion_id: "
            f"{getattr(result, 'ingestion_id', 'unknown')})"
        )

    return [search_memory, store_memory]

File: packages/integrations/synap-eve/src/tools.ts
// Synap tools for eve agents.
//
// Each factory returns an eve `ToolDefinition` (branded by `defineTool`) meant
// to be the default export of a file under `agent/tools/`, where the filename
// becomes the model-facing tool name:
//
//   // agent/tools/synap_search.ts
//   import { createSynapSearchTool } from "@maximem/synap-eve";
//   import { sdk } from "../lib/synap.js";
//   export default createSynapSearchTool({ sdk });
//
// Identity is auto-scoped from the eve session (`ctx.session.auth.principalId`),
// with an explicit `userId` override for unauthenticated channels.
//
// Error policy: reads (search) degrade to an empty result; writes (store) raise
// `SynapIntegrationError` so ingestion outages are observable to the agent.

import { defineTool } from "eve/tools";
import { z } from "zod";

import {
  SynapIntegrationError,
  resolveUserId,
  resolveConversationId,
  type EveSessionLike,
  type SynapSdkLike,
} from "./types.js";

export interface SynapToolOptions {
  /** Configured Synap SDK. */
  sdk: SynapSdkLike;
  /** Explicit Synap user scope. Overrides the eve principal; required on unauthenticated channels. */
  userId?: string;
  /** Optional customer/org scope. */
  customerId?: string;
  /** Explicit conversation id. Defaults to the eve session id. */
  conversationId?: string;
  /** Synap fetch mode ("accurate" default, or "fast"). */
  mode?: string;
}

const searchInput = z.object({
  query: z.string().describe("What to look up in the user's memory"),
  maxResults: z
    .number()
    .int()
    .positive()
    .optional()
    .describe("Maximum number of results (default 10)"),
});

const storeInput = z.object({
  content: z.string().describe("The fact, preference, or note to persist"),
  metadata: z
    .record(z.string(), z.unknown())
    .optional()
    .describe("Optional metadata to attach"),
});

/**
 * Read tool: search the user's Synap memory for context relevant to a query.
 * Degrades to `{ available: false }` on missing identity or SDK failure — a
 * recall miss never aborts the turn.
 */
export function createSynapSearchTool(options: SynapToolOptions) {
  const { sdk } = options;
  if (!sdk) throw new TypeError("createSynapSearchTool requires a non-null sdk");

  return defineTool({
    description:
      "Search the user's long-term Synap memory (facts, preferences, " +
      "episodes, emotions, temporal events) for context relevant to a " +
      "query. Use when you need background about the user that is not in " +
      "the current conversation.",
    inputSchema: searchInput,
    outputSchema: z.object({
      formattedContext: z.string(),
      available: z.boolean(),
    }),
    async execute(input, ctx) {
      const query = (input?.query ?? "").trim();
      const maxResults = input?.maxResults ?? 10;
      const userId = resolveUserId(options.userId, ctx as EveSessionLike);
      if (!query || !userId) {
        return { formattedContext: "", available: false };
      }
      try {
        const response = await sdk.fetch({
          conversation_id: resolveConversationId(options.conversationId, ctx as EveSessionLike) ?? null,
          user_id: userId,
          customer_id: options.customerId || null,
          search_query: [query],
          max_results: maxResults,
          mode: options.mode ?? "accurate",
          include_conversation_context: false,
        });
        const text = (response.formatted_context ?? "").trim();
        return { formattedContext: text, available: !!text };
      } catch (err) {
        console.error(`[synap-eve] synap_search failed userId=${userId}:`, err);
        return { formattedContext: "", available: false };
      }
    },
  });
}

/**
 * Write tool: persist an explicit fact/preference/note to Synap for future
 * recall. Raises `SynapIntegrationError` on failure (writes never silently
 * drop) — eve surfaces the error result to the model.
 */
export function createSynapStoreTool(options: SynapToolOptions) {
  const { sdk } = options;
  if (!sdk) throw new TypeError("createSynapStoreTool requires a non-null sdk");

  return defineTool({
    description:
      "Persist an explicit fact, preference, or note to the user's Synap " +
      "memory for future recall. Call this when the user shares something " +
      "worth remembering across sessions.",
    inputSchema: storeInput,
    outputSchema: z.object({
      recorded: z.boolean(),
      ingestionId: z.string().optional(),
    }),
    async execute(input, ctx) {
      const content = (input?.content ?? "").trim();
      const userId = resolveUserId(options.userId, ctx as EveSessionLike);
      if (!content) {
        throw new SynapIntegrationError("synap_store", "missing content");
      }
      if (!userId) {
        throw new SynapIntegrationError(
          "synap_store",
          "no user scope (authenticate the channel or pass userId)",
        );
      }
      const metadata: Record<string, unknown> = { source: "eve", ...(input?.metadata ?? {}) };
      try {
        const result = await sdk.memories.create({
          document: content,
          user_id: userId,
          customer_id: options.customerId || null,
          metadata,
        });
        return { recorded: true, ingestionId: result.ingestion_id ?? "" };
      } catch (err) {
        const msg = err instanceof Error ? err.message : String(err);
        console.error(`[synap-eve] synap_store failed userId=${userId}:`, err);
        throw new SynapIntegrationError("synap_store", msg, err);
      }
    },
  });
}
